external-crates/move/crates/move-compiler/src/parser/syntax.rs

```diff
@@ -1514,21 +1599,227 @@ fn parse_name_exp(context: &mut Context) -> Result<Exp_, Box<Diagnostic>> {
+                    }
+                    Tok::LBrace => {
+                        let (loc, patterns) = with_loc!(
+                            context,
+                            parse_comma_list(
+                                context,
+                                Tok::LBrace,
+                                Tok::RBrace,
+                                parse_field_pattern,
+                                "a field pattern",
+                            )?
+                        );
+                        FieldConstructor(name_access_chain, ty_args, sp(loc, patterns))
+                    }
+                    _ => Name(mut_, name_access_chain, ty_args),
+                }
+            }),
+            _ => {
+                if let Some(value) = maybe_parse_value(context)? {
+                    Ok(sp(value.loc, Literal(value)))
+                } else {
+                    Err(Box::new(diag!(
+                        Syntax::UnexpectedToken,
+                        (context.tokens.current_token_loc(), INVALID_PAT_ERROR_MSG)
+                    )))
+                }
+            }
+        }
+    }
+
+    fn parse_positional_field_pattern(
+        context: &mut Context,
+    ) -> Result<Ellipsis<MatchPattern>, Box<Diagnostic>> {
+        if let Some(loc) = parse_ellipsis_opt(context)? {
+            return Ok(Ellipsis::Ellipsis(loc));
+        }
+
+        parse_match_pattern(context).map(Ellipsis::Binder)
+    }
+
+    fn parse_field_pattern(
+        context: &mut Context,
+    ) -> Result<Ellipsis<(Field, MatchPattern)>, Box<Diagnostic>> {
+        const INVALID_MUT_ERROR_MSG: &str = "'mut' modifier can only be used on variable bindings";
+
+        if let Some(loc) = parse_ellipsis_opt(context)? {
+            return Ok(Ellipsis::Ellipsis(loc));
+        }
+
+        let mut_ = parse_mut_opt(context)?;
+        let field = parse_field(context)?;
+        let pattern = if match_token(context.tokens, Tok::Colon)? {
+            if let Some(loc) = mut_ {
+                return Err(Box::new(diag!(
+                    Syntax::UnexpectedToken,
+                    (loc, INVALID_MUT_ERROR_MSG)
+                )));
+            }
+            parse_match_pattern(context)?
+        } else {
+            sp(
+                field.loc(),
+                Name(mut_, sp(field.loc(), NameAccessChain_::One(field.0)), None),
+            )
+        };
+        Ok(Ellipsis::Binder((field, pattern)))
+    }
+
+    fn parse_at_pattern(context: &mut Context) -> Result<MatchPattern, Box<Diagnostic>> {
+        match context.tokens.peek() {
+            Tok::Identifier if context.tokens.lookahead() == Ok(Tok::AtSign) => {
+                if context.tokens.content() == "_" {
+                    Err(Box::new(diag!(
+                        Syntax::UnexpectedToken,
+                        (context.tokens.current_token_loc(), WILDCARD_AT_ERROR_MSG)
+                    )))
```

Comment:
This feels a bit strange to have as a parser error. For all existing wildcard/underscore cases, we parse it as an identifier and then clean it up during expansion. I feel like we should do that to be consistent here